### PR TITLE
Update persyst.py

### LINKED ISSUE #8292: persyst annotations error
**READ THIS FIRST:** If you are having trouble getting MNE-Python to work with
your own data, you should probably ask for help on our
[Gitter channel](https://gitter.im/mne-tools/mne-python) or on our
[mailing list](https://mail.nmr.mgh.harvard.edu/mailman/listinfo/mne_analysis)
before opening an issue; you will get help faster that way. See
[the FAQ entry on filing bug reports](https://mne.tools/dev/overview/faq.html#i-think-i-found-a-bug-what-do-i-do)
for more guidance. If you're pretty sure your problem is a bug in MNE-Python,
**delete this section** and fill in the headings below, replacing the
placeholder text with your own responses. Surround any code samples with triple
backticks above and below the code block, or put code in a
[public gist](https://gist.github.com) and link to it in the issue.

#### Describe the bug
The persyst raw file function is able to open the persyst files and annotations with no issues. The issue I am having. For my particular use case we need to get all annotations regardless if they are repeated. When some of the EEG systems generate annoations they use xlevents or xlspikes and sometimes they can be multiple of these in the files. however the parser is only taking unique annotations. So if a human annotator or machine annotator uses same annotations then the parser is not picking up all of the annotations but rather unique names or the first one with that name. 

#### Steps to reproduce
Open any lay/dat files that have annotations with different times and same name/description

[MWE]: https://en.wikipedia.org/wiki/Minimal_Working_Example
[built-in datasets]: https://mne.tools/dev/overview/datasets_index.html
[sample]: https://mne.tools/dev/overview/datasets_index.html#sample

#### Expected results
The expected behavior is that the parser should be able to pick up all annotations even if they have the same name. 

#### Actual results
only picks up the first annotation with that name. 

#### Additional information
If I have time I can try to submit a suggestion on how to fix it.

### PR DESCRIPTION
**Pull request by infinitejest**
Fixes issue that annotation parser is not able to parse comments with same description

#### Reference issue
Fixes #8292 

#### What does this implement/fix?
Persyst annotation parser of Lay files is not able to parse when annotations have same description over different annotation time and duration. This fix creates lists of tuples of annotation with same description and then creates an annotation instance of each individual annotation including repeated instances. 

#### Additional information
OF note when try to create events from annotation objects events are not able to recognize multiple instances of annotation objects with same description.

### CODE REVIEW

**Comment by hoechenberger** [comment]
Thanks for your contribution @infinitejest! Do you think you could add a test, too?

**Comment by infinitejest** [comment]
That is the one thing neuroscientists hate to do but I'll try.

Andres

**Comment by hoechenberger** [comment]
Let us know if you get stuck.

**Comment by hoechenberger** [comment]
@nfinitejest How are things? :)

**Comment by infinitejest** [comment]
Have not had time , but probably by the end of the week I will give those
tests a try

**Comment by hoechenberger** [comment]
Great! Do not hesitate to ask for help should you get stuck or lost in the code 😁

**Comment by infinitejest** [comment]
> Great! Do not hesitate to ask for help should you get stuck or lost in the code 😁

I was looking into test_persyst but there is no tests for the annotations. How do you tests that annotations are done correctly ?

**Comment by agramfort** [comment]
yes we need a test. Lesson learnt. We add something without a test and 3
weeks later a pb occurs :)

>

**Comment by infinitejest** [comment]
Well now that I am at the task what do you think is important to test in th
annotations ?

Andres Rodriguez

**Comment by agramfort** [comment]
a test should be added to avoid a regression. Basically add a test that
would fail on master
and that now works with your PR.

simple :)

>

**Comment by infinitejest** [comment]
I guess this is more of a functional test because in order to get that to
happen the only difference is the output of the main branch and the PR. Do
you have any examples of annotations where there are repeated descriptions.
I supposed we can test to an expected output ? will that be a good test ?


Andres Rodriguez

[image: Mailtrack]
<https://mailtrack.io?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality5&>
Sender
notified by
Mailtrack
<https://mailtrack.io?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality5&>
10/07/20,
10:31:46 AM

**Comment by agramfort** [comment]
yes I think we would need a file

>

**Comment by adam2392** [comment]
> Sorry for my slow progress. So what I did was I went a head and went to our clinical system and created a fake patient with no EEG tracing but created duplicate labels. Now that I have that. I wanted to ask you how do I go about setting up that in the Repository so you have that file and then set it up for the testing environment? 

Hi @infinitejest sorry for the slow reply. Didn't realize this fell through the cracks. If you create a short (2-3 second) file, of the file you have, you can make a PR into `mne-testing-data` directory, which I believe (@agramfort and/or @larsoner  would know better). Since there is a Persyst file already there, you can perhaps add another "subject", and perhaps add a note saying what that file's intended to test for?

Then when that PR is successful, someone will merge it in, generate a new version and a hashsum, so that way `mne-python` can be updated to use the new testing data. Then you can test with that new version that will contain your proposed test file.

https://github.com/mne-tools/mne-testing-data

**Comment by infinitejest** [comment]
Thanks, I'll do that next week when I have more time.

I am running into new problem. When I run the persyst file script it works
fine on windows but not on Linux. Do you know if this script is looking
into a metadata inside the lay/day files that may cause to work in one but
not the other. I am imagining that when lay/day are created it has some
data that connects the two.that through an error on Linux.

Andres Rodriguez

**Comment by infinitejest** [comment]
> > Sorry for my slow progress. So what I did was I went a head and went to our clinical system and created a fake patient with no EEG tracing but created duplicate labels. Now that I have that. I wanted to ask you how do I go about setting up that in the Repository so you have that file and then set it up for the testing environment?
> 
> Hi @infinitejest sorry for the slow reply. Didn't realize this fell through the cracks. If you create a short (2-3 second) file, of the file you have, you can make a PR into `mne-testing-data` directory, which I believe (@agramfort and/or @larsoner would know better). Since there is a Persyst file already there, you can perhaps add another "subject", and perhaps add a note saying what that file's intended to test for?
> 
> Then when that PR is successful, someone will merge it in, generate a new version and a hashsum, so that way `mne-python` can be updated to use the new testing data. Then you can test with that new version that will contain your proposed test file.
> 
> https://github.com/mne-tools/mne-testing-data

Where you able to incorporate my code into main ?

**Comment by hoechenberger** [comment]
Hello @infinitejest, it appears this code has not yet been merged into `main`. Kindly pinging @adam2392, could you have a look what needs to be done so we can move ahead?

**Comment by infinitejest** [comment]
Thanks I will upload another pull request. The persyst.py has a bug that if annotations have a comma it will break the script. I made them skip comments like that for now. Do you want to put in separate pull request ?

**Comment by adam2392** [comment]
Hi @infinitejest you mentioned a test dataset that was able to demonstrate the issue. Do you mind uploading it here, or providing a Dropbox link? Once we get that, then we can proceed.

Any code changes that improves the Persyst reader are greatly appreciated! You can just stick them in the current PR. 

@hoechenberger I can go ahead and make a PR into the mine-testing-data to make the master branch break, and then go ahead and review @infinitejest ’s PR here.

**Comment by infinitejest** [comment]
Yes that is easier. I didn't get the link.

Andres Rodriguez

**Comment by adam2392** [comment]
> Yes that is easier. I didn't get the link. Andres Rodriguez

Sorry I meant are you able to upload a link with the data you mentioned here: https://github.com/mne-tools/mne-python/pull/8311#issuecomment-707757614 

If the “test patient” is small enough, you can just compress and upload via github as a zip file, or if you put it up on Dropbox/GDrive, I can download it asap.

Just for the sake of having a test dataset to test these changes against.

File: mne/io/persyst/persyst.py [hunk 1]
```diff
@@ -108,8 +111,29 @@ def __init__(self, fname, preload=False, verbose=None):
             # Patient (All optional)
             elif section == 'patient':
                 patient_dict[key] = val
+
```
Comment (adam2392) [inline mne/io/persyst/persyst.py:114]: I think unnecessary spaces will break flake8

**Comment by hoechenberger** [comment]
> @hoechenberger I can go ahead and make a PR into the mine-testing-data to make the master branch break, and then go ahead and review @infinitejest ’s PR here.

Yep, sounds good. Be sure to follow the [readme](https://github.com/mne-tools/mne-testing-data#mne-testing-data), as the process consists of numerous steps.

**Comment by infinitejest** [comment]
Here you go: 
I think the easiest way to test this is by checking the number of annotations. The main branch will find 8 and my PR should do 12. 


Link to files
https://emory.box.com/s/2oe1jf68q2k1tl55nrjmj1i5549bow40

**Comment by adam2392** [comment]
Hi @infinitejest I just clicked on the link and I got this:

Screen Shot 2021-01-28 at 11 38 27 AM

Sorry for the wait, but basically we need a test dataset, or at least a script that modifies the existing test dataset in unit tests that break functionality as you observed. Then it's easier to clean up the code here and merge it in and add unit tests. (https://github.com/mne-tools/mne-python/pull/8311#issuecomment-768937468)

**Comment by infinitejest** [comment]
I was sharing my test lay/dat files. Maybe just to give you the annotations table is enough ?

**Comment by adam2392** [comment]
I don't have write access to this PR. I just fired up a new one at: https://github.com/mne-tools/mne-python/pull/8806

**Pull request closed without merging**